### ISSUE TITLE
starmap

### ISSUE THREAD

**Issue by Veedrac** [rust-itertools/itertools#139]
It'd be nice to have a `starmap` method. The main aim is to skip having to write closures after things like `zip`. If you're not familiar with it from Python, a `starmap` function unpacks the tuples in the input iterable into the function being mapped.

A simple but unstable implementation is

```
fn starmap<F>(self, f: F) -> iter::Map<Self, FnZipper<F>>
    where Self: Sized, F: Fn<Self::Item>
{
    self.map(FnZipper(f))
}
```

where

```
pub struct FnZipper<F>(F);

impl<F, Args> FnOnce<(Args,)> for FnZipper<F>
    where F: FnOnce<Args>
{
    type Output = F::Output;
    extern "rust-call" fn call_once(self, (args,): (Args,)) -> Self::Output {
        self.0.call_once(args)
    }
}

impl<F, Args> FnMut<(Args,)> for FnZipper<F>
    where F: FnMut<Args>
{
    extern "rust-call" fn call_mut(&mut self, (args,): (Args,)) -> Self::Output {
        self.0.call_mut(args)
    }
}

// For completeness
impl<F, Args> Fn<(Args,)> for FnZipper<F>
    where F: Fn<Args>
{
    extern "rust-call" fn call(&self, (args,): (Args,)) -> Self::Output {
        self.0.call(args)
    }
}
```

A stable implementation would probably requires macros to handle all the wanted cases, and macros would also let us also support more sophisticated flattening like for

```
xs.zip(ys).zip(zs).starmap(|x, y, z| ...)
```

Thoughts?

**Comment by bluss** [comment]
The unstable impl is neat, but i don't want unstable features the main itertools crate anymore. (I'm going to remove those that are in the next version, or before that if they break before, like .enumerate_from()).

We've got some tuple flattening macros and impls already for `Zip`, so then we just need something that accepts non-tuple closures for tupled iterator elements, that's certainly a nice thing to have.

**Comment by bluss** [comment]
WIP in #140.

It does no flattening, because flattening is already available as part of the n-ary Zip (well, "n-ary" since it's for small n only).

**Comment by nirvana-msu** [comment]
Coming from Python, this seems missing.  It really does make your code more neat when you can just pass a method/function into map-like combinator, instead of having to explicitly write up closures only to unpack a tuple into arguments.

**Comment by Philippe-Cholet** [comment]
**bluss** implemented a draft in #140 but did not merge it mostly arguing that it could probably be done by `map` itself.
I come from Python's itertools which I loved but I am hesitant about this too.
Seeing there is no like (but 1 liking comment) in 7 years, I close this.
